### PR TITLE
Use the built-in tool wslpath

### PR DESCRIPTION
**Pull request by jmerriweather**
Should we use wslpath instead of the regex?

### CODE REVIEW

File: lib/mix/nerves/utils.ex [hunk 1]
```diff
@@ -195,16 +195,16 @@ defmodule Mix.Nerves.Utils do
...
+    # -a    force result to absolute path format
+    # -u    translate from a Windows path to a WSL path (default)
+    # -w    translate from a WSL path to a Windows path
+    # -m    translate from a WSL path to a Windows path, with ‘/’ instead of ‘\\’
 
-    drive_letter =
-      Regex.run(~r/(.*?):\\/, win_path)
-      |> Enum.at(1)
-      |> String.downcase()
+    win_path = System.cmd("wslpath", ["-w -a", file])
```
Comment (fhunleth) [inline lib/mix/nerves/utils.ex]: Did you test this? I'd think that this line should be `{win_path, 0} = System.cmd("wslpath", ["-w", "-a", file])`.

**Comment by michaelkschmidt** [comment]
Note that `wslpath` is relatively new--this will break compatibility with Creators Update and Fall Creators Update

**Comment by fhunleth** [comment]
It sounds like we should hold off on changing things right now.

**Comment by jmerriweather** [comment]
Yeah sorry multiple errors in this PR. I was also thinking about this and wslpath can give a "Result not representable" when the path is not available from the Windows side: `> wslpath: <filepath>: Result not representable`. 

wslpath is available from Windows 10 1803 which was released April 2018. Windows 10 1809 is about to come out, how many versions of Windows are you planning to support?

**Comment by jmerriweather** [comment]
I've fixed **and tested** this branch. I've also added a feature:

- When using nerves in WSL from a path Windows can access, don't copy the firmware

If you are happy to support only the latest version (soon to be latest two versions) of Windows 10, let me know and I'll create a new PR.

**Comment by fhunleth** [comment]
@jmerriweather Thanks for fixing and updating. I reopened the PR to continue discussion.

**Comment by jmerriweather** [comment]
As nerves support for _Windows Subsystem for Linux_ is experimental I don't see an issue with having an expectation that users will need the latest version of Windows 10

**Comment by michaelkschmidt** [comment]
Please don’t get me wrong @jmerriweather, I think this is good stuff.  I’m all for adding support for using wslpath but there needs to be a fallback for when it’s not there. 

Windows support has always been driven by Corporate users. You can’t drop support for a version of Windows that is less than a year old—it takes months and months for big IT to approve releases when they come out

**Comment by jmerriweather** [comment]
@michaelkschmidt all good, it's good to talk this through and work out the best approach :)

Microsoft recently increased support for Enterprise and Education to 30 months from release. Luckily WSL isn't available in the LTSB version which has a 10 year support cycle. If full Windows support is required we'll need to provide an alternative code path until April 2020.

If the project wants to provide full support for Windows 10 I could rework the PR to fail back to the old code path?

**Comment by michaelkschmidt** [comment]
I don't think we need quite that long--I'm just advocating for 1 year-ish, so I can still use it 😉 

I think `wslpath` will work with `System.find_executable()`.  Can we fall back to the existing methods if its not there?  I think you would only need to touch `get_wsl_paths(file)` and `is_wsl_path_accessible?(file)`.  The former can fall back to the existing calls, and the latter can just default to false.  This would retain the current behavior, while giving modern WSL users a much better experience

**Comment by fhunleth** [comment]
I have mixed feelings about this. Maybe you two can help out.

1. Does Windows have official rules for converting paths between WSL and Windows?
2. Are those rules at risk of changing? (I'm sure anything possible, but I'm wondering if Microsoft has a note somewhere that says the mapping is a hack and not to rely on it or something like that)

What I'm getting at is that it seems like `wslpath` reduces the risk of the WSL/Windows path conversion rules changing on us. That's a good thing since it's one less thing that can go wrong.

If, however, Windows has an unambiguous set of rules for the conversion and if they're not unwieldy, would it be too much work to code them up? I assume that the existing regular expression is close and if a few unit tests were added, would that give everyone confidence in it?

I understand the arguments about Windows versions and compatibility, so don't think that I'm ignoring it.

**Comment by jmerriweather** [comment]
I've tested the above commit with Windows 10 1803, @michaelkschmidt could you test this on a previous build of Windows 10?

**Comment by jmerriweather** [comment]
@fhunleth I don't know of any official rules, you can actually change the root mount path of windows using `wsl.conf` as outlined here: https://blogs.msdn.microsoft.com/commandline/2018/02/07/automatically-configuring-wsl/ 

I feel that using wslpath is the safest option as Windows injects the binary it into the Linux distribution. I'm hoping my latest commit is a good compromise

File: lib/mix/tasks/firmware.burn.ex [hunk 2]
```diff
@@ -63,9 +63,14 @@ defmodule Mix.Tasks.Firmware.Burn do
     # Create a temporary .fw file that fwup.exe is able to access
```
Comment (fhunleth) [inline lib/mix/tasks/firmware.burn.ex]: The "create a temporary .fw file" comment should be moved down to line 69.5.

File: lib/mix/nerves/utils.ex [hunk 3]
```diff
@@ -194,18 +194,46 @@ defmodule Mix.Nerves.Utils do
     end
   end
 
-  def get_wsl_paths(file) do
-    {win_path, 0} = System.cmd("cmd.exe", ["/c", "cd"])
-    win_path = String.trim(win_path) <> "\\#{file}"
+  def is_wslpath_available?() do
```
Comment (fhunleth) [inline lib/mix/nerves/utils.ex]: There's an Elixir style to only put `is_` in functions that can be guards, so this should be named `wslpath_available?` or `has_wslpath?`

File: lib/mix/tasks/firmware.burn.ex (hunk 2 above)
Comment (jmerriweather) [reply lib/mix/tasks/firmware.burn.ex]: ah yes, i'll move that down

File: lib/mix/nerves/utils.ex (hunk 3 above)
Comment (jmerriweather) [reply lib/mix/nerves/utils.ex]: I wasn't aware of that. I think has_wslpath? might be a better name

**Comment by jmerriweather** [comment]
My pleasure, glad we could come to a solution that works for everyone :smiley:

File: lib/mix/nerves/utils.ex [hunk 4]
```diff
@@ -194,18 +194,46 @@ defmodule Mix.Nerves.Utils do
...
-    win_path = String.trim(win_path) <> "\\#{file}"
+  def has_wslpath?() do
+    System.find_executable("wslpath") != nil
+  end
 
-    drive_letter =
-      Regex.run(~r/(.*?):\\/, win_path)
-      |> Enum.at(1)
-      |> String.downcase()
+  def is_wsl_path_accessible?(file) do
```
Comment (fhunleth) [inline lib/mix/nerves/utils.ex]: One more function to remove the `is_` prefix. Fwiw, here's the logic: https://github.com/elixir-lang/elixir/blob/master/lib/elixir/pages/Naming%20Conventions.md#is_-prefix-is_foo

**Pull request merged**